### PR TITLE
Darwin: Fix some more ambiguous Swift symbols.

### PR DESCRIPTION
**Pull request by ksperling-apple**
The following were still ambiguous in Swift:
- MTRCommissioningParameters.csrNonce
- MTROTAProviderDelegate.handleBDXTransferSessionBegin
- MTROTAProviderDelegate.handleBDXQuery

Use MTR_SWIFT_DISFAVORED_OVERLOAD instead of NS_UNAVAILABLE_FOR_SWIFT and introduce MTR_DEPRECATED that makes use of it to make Swift prefer non-deprecated symbols where there is ambiguity due to the old and new ObcJ name being imported as the same name in Swift.

Also avoid some typedefs for single-use completion block types since they add little value but carry a cost in terms of maintaining source compatibility going forward.

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h [hunk 1]
```diff
@@ -64,12 +59,13 @@ typedef void (^MTRApplyUpdateRequestCompletionHandlerDeprecated)(
...
                        completion:(MTRQueryImageCompletionHandler)completion
     API_AVAILABLE(ios(16.4), macos(13.3), watchos(9.4), tvos(16.4));
+
 - (void)handleQueryImageForNodeID:(NSNumber *)nodeID
                        controller:(MTRDeviceController *)controller
                            params:(MTROtaSoftwareUpdateProviderClusterQueryImageParams *)params
-                completionHandler:(MTRQueryImageCompletionHandlerDeprecated)completionHandler
-    API_DEPRECATED(
-        "Please use the selector ending in completion:", ios(16.1, 16.4), macos(13.0, 13.3), watchos(9.1, 9.4), tvos(16.1, 16.4));
+                completionHandler:(MTRQueryImageCompletionHandler)completionHandler
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:66]: This doesn't look right.  The types of the `data` arg are different for `MTRQueryImageCompletionHandler` and `MTRQueryImageCompletionHandlerDeprecated` ("Ota" vs "OTA" in the struct name), so code written to call the deprecated version of `handleQueryImageForNodeID` won't compile after this change, right?

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h [hunk 2]
```diff
@@ -85,12 +81,13 @@ typedef void (^MTRApplyUpdateRequestCompletionHandlerDeprecated)(
...
                                completion:(MTRApplyUpdateRequestCompletionHandler)completion
     API_AVAILABLE(ios(16.4), macos(13.3), watchos(9.4), tvos(16.4));
+
 - (void)handleApplyUpdateRequestForNodeID:(NSNumber *)nodeID
                                controller:(MTRDeviceController *)controller
                                    params:(MTROtaSoftwareUpdateProviderClusterApplyUpdateRequestParams *)params
-                        completionHandler:(MTRApplyUpdateRequestCompletionHandlerDeprecated)completionHandler
-    API_DEPRECATED(
-        "Please use the selector ending in completion:", ios(16.1, 16.4), macos(13.0, 13.3), watchos(9.1, 9.4), tvos(16.1, 16.4));
+                        completionHandler:(MTRApplyUpdateRequestCompletionHandler)completionHandler
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:88]: Again, the type of the completion is being changed here and probably shouldn't be.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h [hunk 3]
```diff
@@ -105,12 +102,13 @@ typedef void (^MTRApplyUpdateRequestCompletionHandlerDeprecated)(
...
                                 completion:(MTRStatusCompletion)completion
     API_AVAILABLE(ios(16.4), macos(13.3), watchos(9.4), tvos(16.4));
+
 - (void)handleNotifyUpdateAppliedForNodeID:(NSNumber *)nodeID
                                 controller:(MTRDeviceController *)controller
                                     params:(MTROtaSoftwareUpdateProviderClusterNotifyUpdateAppliedParams *)params
                          completionHandler:(StatusCompletion)completionHandler
-    API_DEPRECATED(
-        "Please use the selector ending in completion:", ios(16.1, 16.4), macos(13.0, 13.3), watchos(9.1, 9.4), tvos(16.1, 16.4));
+    MTR_DEPRECATED_WITH_REPLACEMENT("handleNotifyUpdateAppliedForNodeID:controller:params:completion:", ios(16.1, 16.4),
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:110]: Question about the semantics of MTR_DEPRECATED_WITH_REPLACEMENT: The other selector has a different type for the `params` argument.  Does that still make it a WITH_REPLACEMENT case?

File: src/darwin/Framework/CHIP/Matter.h [hunk 4]
```diff
@@ -17,14 +17,6 @@
 
 #import <Foundation/Foundation.h>
 
-#ifndef MTR_NEWLY_DEPRECATED
-#define MTR_NEWLY_DEPRECATED(message)
-#endif
-
-#ifndef MTR_NEWLY_AVAILABLE
-#define MTR_NEWLY_AVAILABLE
-#endif
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/Matter.h:26]: Why is this being removed?

Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/Matter.h:26]: Moved to `MTRDefines.h`, leaving `Matter.h` as a pure umbrella header.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h (hunk 1 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:66]: Hm you may be right. However the way the existing type has been renamed to `...Deprecated` wasn't source compatible either. For compatibility purposes just defining the block types inline in the method signature would actually be easier to deal with.

Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:66]: > However the way the existing type has been renamed to ...Deprecated wasn't source compatible either

It's not source compatible if people actually use the typedef types.  The change here is not source compatible if the function is called at all, right?

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h (hunk 3 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:102]: I think in this case it's fine because `MTROtaSoftwareUpdateProviderClusterNotifyUpdateAppliedParams` is also meant to be replaced by the corresponding `MTROTA...` class.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h (hunk 1 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegate.h]: You're right, I've reverted this to the original block types (minus the typedef) and left the code as is. I think the implementation could just cast the block instead of wrap it, but I think it's probably best not to mess with it for this PR.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h [hunk 5]
```diff
@@ -16,26 +16,13 @@
  */
 
 #import <Foundation/Foundation.h>
-#import <Matter/Matter.h>
+#import <Matter/MTRCluster.h>
+#import <Matter/MTRCommandPayloadsObjc.h>
+#import <Matter/MTRDefines.h>
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:21]: Why this change?

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
Fast-tracking platform-specific change with platform owner review.

File: src/darwin/Framework/CHIP/MTROTAProviderDelegate.h (hunk 5 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTROTAProviderDelegate.h:21]: The umbrella header imports all public headers, so public headers shouldn't in turn import the umbrella header, but the actual specific other public headers that they need.
